### PR TITLE
Fix how intersected bounds are added to Constraint

### PR DESCRIPTION
**Pull request by dwijnand**

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/ConstraintHandling.scala [hunk 1]
```diff
@@ -864,6 +864,8 @@ trait ConstraintHandling {
       try bound match
         case bound: TypeParamRef if constraint contains bound =>
           addParamBound(bound)
+        case bound: TypeVar if constraint.contains(bound) =>
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/ConstraintHandling.scala:867]: The comment above `addConstraint` says that `bound` should not be a TypeVar, so it looks like a precondition is being violated if this is needed.

Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/core/ConstraintHandling.scala:867]: Ah, thank you.  I force pushed a subset of the commit, to check it didn't independently break anything.  The TypeVar I'm finding is nested in a bound AndType, not top level.

File: compiler/src/dotty/tools/dotc/core/ConstraintHandling.scala [hunk 2]
```diff
@@ -861,12 +861,24 @@ trait ConstraintHandling {
...
-      try bound match
+
+      def contains(bound: Type) = bound match
+        case tpr: TypeParamRef => constraint.contains(tpr)
+        case tv: TypeVar       => constraint.contains(tv)
+        case _                 => false
+
+      def add(bound: Type): Boolean = bound match
+        case AndType(bound1, bound2) if contains(bound1) ^ contains(bound2) =>
+          add(bound1) && add(bound2)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/ConstraintHandling.scala:872]: I don't understand what's going on here but it seems incorrect. According to the documentation comment this case can only be hit if `bound` is a lower bound, so we're trying to add the following constraint:
```
param >: bound1 & bound2
```
And this case splits this into two two constraints:
```
param >: bound1
param >: bound2
```
Which is equivalent to:
```
param >: bound1 | bound2
```
So we are over-constraining which can prevent some possible solutions (I can't instantiate `param` to `bound1` or `bound2` or their intersection anymore, unless `bound1 =:= bound2`).

**Review by smarter — CHANGES_REQUESTED** [review]

File: compiler/src/dotty/tools/dotc/core/ConstraintHandling.scala (hunk 2 above)
Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/core/ConstraintHandling.scala:872]: No, I'm splitting `P >: B & Q`, where `P` and `Q` are params and `B` isn't, into `P >: B` constraint and `Q < P` ordering, which allows for `P =:= Q` unification to go through.

Comment (smarter) [reply compiler/src/dotty/tools/dotc/core/ConstraintHandling.scala:872]: This still leads to overconstraining, here's an example which compiled before this change but not after:
```scala
class LT[-X, Y]
object LT {
 given lt[X <: Y, Y]: LT[X, Y] = ???
}

trait B

def foo[Q, P](using LT[B & Q, P])(p: P, q: Q): P = p

def test(b: B, any: Any): B =
  foo(b, any)
```
Setting `P =:= Q` means that the method is forced to return `Any`, even though the only thing we specified is that `B & Any <: B`

**Review by smarter — CHANGES_REQUESTED** [review]

**Comment by dwijnand** [comment]
I can't remember what approach I was taking before going down this split-the-bound approach.  Isn't it still a problem that the bounds are recursive?  But the issue it's fixing is kind of niche, so I guess I can leave this for now.

**Pull request closed without merging**